lib/Dialect/Substrait/IR/Substrait.cpp

```diff
@@ -51,6 +51,27 @@ void SubstraitDialect::initialize() {
 namespace mlir {
 namespace substrait {
 
+LogicalResult
+CrossOp::inferReturnTypes(MLIRContext *context, std::optional<Location> loc,
+                          ValueRange operands, DictionaryAttr attributes,
+                          OpaqueProperties properties, RegionRange regions,
+                          llvm::SmallVectorImpl<Type> &inferredReturnTypes) {
+  Value leftInput = operands[0];
+  Value rightInput = operands[0];
+
+  TypeRange leftFieldTypes = leftInput.getType().cast<TupleType>().getTypes();
+  TypeRange rightFieldTypes = rightInput.getType().cast<TupleType>().getTypes();
+
+  llvm::SmallVector<mlir::Type> fieldTypes;
+  fieldTypes.append(leftFieldTypes.begin(), rightFieldTypes.end());
+  fieldTypes.append(rightFieldTypes.begin(), rightFieldTypes.end());
+  auto resultType = TupleType::get(context, fieldTypes);
```

Comment:
OOC why use a tuple here vs multiple return values?

Reply:
I am not using multiple return values because the op really only returns one thing.

However, I do have taken a short-cut here (and everywhere): these ops don't return a single *tuple*, the return a single *relation*. At some point, I should define a `SubstraitRelation` type that, like `Tuple`, will have possibly nested fields.

This is a real necessity, not just for esthetics: Substrait allows a *double* nesting of arrays and tuples (e.g., an array of structs of structs of arrays of atomics) and I need constraints for which atomic values are allowed, which I can't express with `Tuple`s.

I have not done this initially because I `Tuple` provided a sufficiently close alternative and I wanted to make progress with other things first. I have a TODO in `SubstraitTypes.td`, and we can start working on that whenever we feel the time has come:

```
// TODO(ingomueller): Transform this into a proper relation type.
```